api/bibliotheca.py

```diff
@@ -1413,96 +1461,221 @@ def timestamp(self):
-        # NB: If the timespan is too long, we might yet end up in Backlog Mode.
-        if progress.counter == recent_mode_counter_value:
-            raise_on_no_events = True
-            timespan_to_check = recent_mode_timespan
+        # Record the addition of this title to the collection. Do this
+        # even if we've heard about the book before, through some
+        # other monitor. That's because this is the only Bibliotheca
+        # monitor that issues TITLE_ADD events.
+        #
+        # We know approximately when the license was purchased --
+        # potentially a long time ago -- since `start_time` is
+        # provided.
+        license_pool.collect_analytics_event(
+            self.analytics, CirculationEvent.DISTRIBUTOR_TITLE_ADD,
+            purchase_time, 0, 1
+        )
+        return license_pool
+
+
+class BibliothecaEventMonitor(BibliothecaTimelineMonitor):
+
+    """Register CirculationEvents for Bibliotheca titles.
+
+    When run, this monitor will look at recent events as a way of keeping
+    the local collection up to date.
+
+    Although useful in everyday situations, the events endpoint will
+    not always give you all the events:
+
+    * Any given call to the events endpoint will return at most
+      100-150 events. If there is a particularly busy 5-minute
+      stretch, events will be lost.
+
+    * The Bibliotheca API has, in the past, gone into a state where
+      this endpoint returns an empty list of events rather than an
+      error message.
+
+    Fortunately, we have the BibliothecaPurchaseMonitor to keep track
+    of new license purchases, and the BibliothecaCirculationSweep to
+    keep up to date on books we already know about. If the
+    BibliothecaEventMonitor stopped working completely, the rest of
+    the system would continue to work, but circulation data would
+    always be a few hours out of date.
+
+    Thus, running the BibliothecaEventMonitor alongside the other two
+    Bibliotheca monitors ensures that circulation data is kept up to date
+    in near-real-time with good, but not perfect, consistency.
+    """
 
-            if (cutoff - start) > timespan_to_check:
-                raise_on_no_events = False
-                progress.counter = backlog_mode_counter_value
-                timespan_to_check = backlog_mode_timespan
+    SERVICE_NAME = "Bibliotheca Event Monitor"
 
+    def catch_up_from(self, start, cutoff, progress):
+        self.log.info(
+            "Requesting events between %s and %s",
+            start.strftime(self.LOG_DATE_FORMAT),
+            cutoff.strftime(self.LOG_DATE_FORMAT)
+        )
+        events_handled = 0
+
+        # Since we'll never get more than about 100 events from a
+        # single API call, slice the timespan into relatively small
+        # chunks.
         for slice_start, slice_cutoff, full_slice in self.slice_timespan(
-            start, cutoff, timespan_to_check
+            start, cutoff, timedelta(minutes=5)
```

Comment:
Since it's no longer essential that we get all the events from this monitor, we can go back to a hard-coded delta that will work pretty well most of the time.